src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Reference.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+
+namespace System.Text.RegularExpressions
+{
+    /// <summary>
+    /// Used to cache a weak reference in a threadsafe way
+    /// </summary>
+    internal sealed class SharedReference
+    {
+        private WeakReference _ref = new WeakReference(null);
+        private int _locked;
+
+        /// <summary>
+        /// Return an object from a weakref, protected by a lock.
+        /// 
+        /// If the exclusive lock can't be obtained, null is returned.
+        /// Note that _ref.Target is referenced only under the protection of the lock. (Is this necessary?)
+        /// </summary>
+        internal object Get()
+        {
+            if (0 == Interlocked.Exchange(ref _locked, 1))
+            {
+                object obj = _ref.Target;
+                _locked = 0;
+                return obj;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Suggest an object into a weakref, protected by a lock.
+        /// 
+        /// Note that _ref.Target is referenced only under the protection of the lock. (Is this necessary?)
+        /// </summary>
+        /// <param name="obj"></param>
+        internal void Cache(object obj)
+        {
+            if (0 == Interlocked.Exchange(ref _locked, 1))
+            {
+                _ref.Target = obj;
+                _locked = 0;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Used to cache one exclusive runner reference
+    /// </summary>
+    internal sealed class ExclusiveReference
+    {
+        private RegexRunner _ref;
+        private object _obj;
+        private int _locked;
+
+        /// <summary>
+        /// Return an object and grab an exclusive lock.
+        /// 
+        /// If the exclusive lock can't be obtained, null is returned.
+        /// If the object can't be returned, the lock is released.
+        /// </summary>
+        /// <returns></returns>
+        internal object Get()
+        {
+            // try to obtain the lock
+            if (0 == Interlocked.Exchange(ref _locked, 1))
```

Comment:
Why not just use a normal lock?

Reply:
No idea. I think that change was made by @justinvp. Maybe he can answer?

Reply:
I didn't change it. This is how it was when the sources were added to corefx.

https://github.com/dotnet/corefx/commit/cc07fbf1a2fdac12523573ffb2dd273d82542f8d#diff-18063ba409511b10aec57135398e72d3R1056

Reply:
thanks and sorry for the noise!